### ISSUE TITLE
Thoughts on HARK/SHARKFIN timing

### ISSUE THREAD

**Issue by llorracc** [sbenthall/SHARKFin#58]
This is a followup to #57.

@sbenthall had suggested that one way to align the structure of the problem that agents are solving better in HARK to the requirement of AMP to have trading occur every day, would be basically, at the simulation stage, to have 60 different populations of agents each of which would have a different "customary day" of attention out of the 60 days of the quarter.

Having thought it over, I see some problems with that approach:

1. In order to guarantee a large enough number of participants to make the market "thick" (not easily whipsawed by the choices of a single agent) we'd have to make each of those 60 populations as large as we are now making the overall population.  That means 60 times slower.
2. It would not allow a mechanism for having spikes in the number of people trading as a consequence of big moves in the market.

Seb and I also discussed the idea that we align the HARK probability of "participating" in the market with what has heretofore been called "attending" to the market in SHARKFIN.

But let me clarify one thing: Even the people who are not "participating" should update both their beliefs about their risky asset holdings and their expectations.

Furthermore, the expectations updating step should take the following form:
1.  Agents believe the stock market has two "regimes" : High and low volatility
2. If their beliefs are that the market is in the "low volatility" state but actual volatility has been higher than expected in the low volatility state, their beliefs have a _probability_ of switching to the "high volatility" beliefs that is related to the amount by which actual volatility has exceeded the low-vol prediction.  
3. This will allow there to be differing opinions about the stock market.
4. These beliefs should be calibrated, somehow, with data like those in [this forthcoming paper](https://doi.org/10.1016/j.jeconom.2020.11.011)

Suppose that people typically adjust their risky share once a year, so that the probability of updating is 0.25 per quarter.

Taking all of that into account, here's a proposal for what to do.

Suppose the person arrives at the beginning of quarter q with financial balances (measured on the 0th day of the quarter) of b_{q+0d}.

For each person in the population, do 60 separate simulations (corresponding to the 60 days of the quarter) corresponding to them actually receiving their labor income and participating on that day.  That is, for the sim for day 0:
1. market resources m_{q+0d} reflect the draws of labor income shocks for the quarter having been drawn on day 0
2. expectations are updated
4. Contingent on those expectations, the agent makes a new portfolio allocation choice

And there is a separate sim for day 1 (that is, the agent is assumed NOT to have drawn their labor income shocks on day 0 or computed their market resources

1. Perform a simulation in which, despite their expectations that they might participate on only 1 day, each of our simulated agents is simulated as having updated their stock market expectations on day 0, but does not draw their labor income shocks for the quarter, or "participate" in the market, until period 1

and so on.  So the case in which that agent's "participation date" is on the last day of the quarter, they will have been updating their expectations during all of the 58 previous days, but only get their labor income and make their "participation" choice in period 59.

We then calculate the aggregate results that would occur if 100 percent of people just happened stochastically to participate on each day, which means we get to use the full population for simulating the market interactions (calculating capital gains etc).  But we only "keep" one outcome for the particular agent as being the one that actually happened to them.

One advantage of doing things this way is that prepares us for an anticipated improvement in the HARK model that I think needs to be added, in which participation is endogenous:  When people change their beliefs about stock market uncertainty, they also readjust their portfolios.  

This is all probably hard to follow; I will explain more in Monday's call.

**Comment by sbenthall** [comment]
@llorracc Would it be possible to provide a document with the model equations for this?

**Comment by llorracc** [comment]
Don't have time to provide such a document.

It is a pretty straightforward thing to do, though, if you thoroughly understand the epistemology of the portfolio choice model.

**Comment by llorracc** [comment]
If you could provide equations describing exactly how things work now, though, I could edit those as needed.

**Comment by sbenthall** [comment]
The best description of how things work now is in the Sloan whitepaper.

It uses a combination of equations, and then a description of the main simulation loop at the quarterly and daily levels (based on an original specification from John).

**Comment by sbenthall** [comment]
I do not yet understand what you are suggesting about the simulation logic (from "do 60 separate simulations" to the end of the issue).

But I do think i understand the proposal for how agents can have high and low volatility market expecations. @alanlujan91 explained this to me today in terms of Krusell-Smith.

> Furthermore, the expectations updating step should take the following form:

> Agents believe the stock market has two "regimes" : High and low volatility
> If their beliefs are that the market is in the "low volatility" state but actual volatility has been higher than expected in the low volatility state, their beliefs have a probability of switching to the "high volatility" beliefs that is related to the amount by which actual volatility has exceeded the low-vol prediction.
> This will allow there to be differing opinions about the stock market.
> These beliefs should be calibrated, somehow, with data like those in [this forthcoming paper](https://doi.org/10.1016/j.jeconom.2020.11.011)
> Suppose that people typically adjust their risky share once a year, so that the probability of updating is 0.25 per quarter.

I think the way to extend the current code to accommodate this is to start building a different FinaceModel class.

The current FinanceModel class is the class that handles the update of agent beliefs. It uses a formula and the past prices to compute the most up-to-date expectations of the market. When the HARK agents update their expectations, they do so by adopting the most recent predictions of the FinanceModel.

https://github.com/sbenthall/SHARKFin/blob/master/sharkfin/hark_portfolio_agents.py#L360

We need a new FinanceModel class that tracks the market prices and determines whether the model is in either a high or a low volatitility phase.

@nicksawhney we can start planning how to generalize the FinanceModel and build an abstract interface, much as you did with the Market.

**Comment by llorracc** [comment]
Didn't get a chance to explain either Mon or Wed.  Maybe joint call btw you and me and @alanlujan91 would help.  

Are we on for 1:30 tomorrow?  We could ask @alanlujan91 to join us.

**Comment by sbenthall** [comment]
I believe issue #60 captures what we agreed on in the meeting today.
I think that between that issue and #57, these changes reflect all the major differences between the target ECON model and the current design.

I'd like to revisit questions about how to most efficiently compute the simulation after we have the model details hashed out and coded up.

**Comment by sbenthall** [comment]
Moving this to 4.1 to be revisted after #60 is complete.
